src/annotator/integrations/pdf-metadata.js

```diff
@@ -131,41 +131,64 @@ export class PDFMetadata {
    *
    * @return {Promise<Metadata>}
    */
-  getMetadata() {
-    return this._loaded.then(app => {
-      let title = document.title;
-
-      if (
-        app.metadata &&
-        app.metadata.has('dc:title') &&
-        app.metadata.get('dc:title') !== 'Untitled'
-      ) {
-        title = /** @type {string} */ (app.metadata.get('dc:title'));
-      } else if (app.documentInfo && app.documentInfo.Title) {
-        title = app.documentInfo.Title;
-      }
-
-      const link = [{ href: fingerprintToURN(app.pdfDocument.fingerprint) }];
-
-      const url = getPDFURL(app);
-      if (url) {
-        link.push({ href: url });
-      }
-
-      return {
-        title: title,
-        link: link,
-        documentFingerprint: app.pdfDocument.fingerprint,
-      };
-    });
+  async getMetadata() {
+    const app = await this._loaded;
+    const {
+      info: documentInfo,
+      contentDispositionFilename,
+      metadata,
+    } = await app.pdfDocument.getMetadata();
+    const documentFingerprint = app.pdfDocument.fingerprint;
+
+    const url = getPDFURL(app);
+
+    // Return the title metadata embedded in the PDF if available, otherwise
+    // fall back to values from the `Content-Disposition` header or URL.
+    //
+    // PDFs contain two embedded metadata sources, the metadata stream and
+    // the document info dictionary. Per the specification, the metadata stream
+    // is preferred if available.
+    //
+    // This logic is similar to how PDF.js sets `document.title`.
+    let title;
+    if (metadata?.has('dc:title') && metadata.get('dc:title') !== 'Untitled') {
+      title = /** @type {string} */ (metadata.get('dc:title'));
+    } else if (documentInfo?.Title) {
+      title = documentInfo.Title;
+    } else if (contentDispositionFilename) {
+      title = contentDispositionFilename;
+    } else if (url) {
+      title = filenameFromURL(url);
+    } else {
+      title = '';
+    }
+
+    const link = [{ href: fingerprintToURN(documentFingerprint) }];
+    if (url) {
+      link.push({ href: url });
+    }
+
+    return {
```

Comment:
It took me a bit to figure out that the `Metadata` type that this returns is not the new `Metadata` type that was defined in these changes—that's a bit confusing!